modules/dag-l0/src/main/scala/io/constellationnetwork/dag/l0/infrastructure/rewards/DelegatedRewardsDistributor.scala

```diff
@@ -0,0 +1,330 @@
+import io.constellationnetwork.syntax.sortedCollection.sortedSetSyntax
+
+import eu.timepit.refined.auto.autoUnwrap
+import eu.timepit.refined.types.all.PosLong
+import eu.timepit.refined.types.numeric.NonNegLong
+import org.typelevel.log4cats.SelfAwareStructuredLogger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object DelegatedRewardsDistributor {
+
+  def make[F[_]: Sync](
+    rewardsConfig: RewardsConfig
+  ): DelegatedRewardsDistributor[F] = new DelegatedRewardsDistributor[F] {
+
+    /** Calculate the total rewards to mint for this epoch based on appropriate rewards mechanism The calculation follows this priority:
+      *   1. Use preconfigured rewards per epoch mapping for epochs before the emission formula transition 2. Use the emission formula if
+      *      configured and epoch is past the transition point 3. Fallback to the rewards distribution algorithms in the Rewards module
+      */
+    def calculateTotalRewardsToMint(epochProgress: EpochProgress): F[Amount] = {
+
+      implicit val rewardOrdering: Ordering[(EpochProgress, Amount)] =
+        (x: (EpochProgress, Amount), y: (EpochProgress, Amount)) => x._1.compare(y._1)
+
+      val transitionEpoch =
+        rewardsConfig.emissionConfig
+          .flatMap(config => Option(config.yearInitial))
+          .getOrElse(EpochProgress.MaxValue)
+
+      rewardsConfig.rewardsPerEpoch.filter {
+        case (epoch, _) =>
+          epoch.value <= epochProgress.value && epochProgress.value < transitionEpoch.value
+      }.maxOption match {
+        case Some((_, amount)) =>
+          amount.pure[F]
+
+        case None if rewardsConfig.emissionConfig.exists(epochProgress.value >= _.yearInitial.value) =>
+          calculateEmissionRewards(epochProgress, rewardsConfig.emissionConfig.get)
+
+        case _ =>
+          Slf4jLogger
+            .getLogger[F]
+            .warn("No emission config found, using fallback reward pool")
+            .as(rewardsConfig.fallbackRewardPoolPerEpoch)
+      }
+    }
+
+    /** Calculate rewards using the new emission formula with deterministic precision: i(t) = i_target + (i_initial - i_target) * e^(-λ *
+      * (Y_current - Y_initial)) * (P_initial / P_current)^i_impact
+      */
+    private def calculateEmissionRewards(
+      epochProgress: EpochProgress,
+      emConfig: EmissionConfig
+    ): F[Amount] = Sync[F].defer {
+      // Get the proper epoch-based configurations
+      val iTarget = emConfig.iTarget
+      val iInitial = emConfig.iInitial
+      val lambda = emConfig.lambda
+      val epochsPerYear = emConfig.epochsPerYear.value
+      val yearInitialEpoch = emConfig.yearInitial.value.value
+      val secondsPerEpoch = emConfig.secondsPerEpoch.value
+      val totalSupply = emConfig.totalSupply.value
+      val iImpact = emConfig.iImpact
+
+      if (emConfig.dagPrices.values.isEmpty) {
+        Slf4jLogger.getLogger[F].error("Empty DAG price configuration").as(Amount(NonNegLong.unsafeFrom(0L)))
```

Comment:
Wondering here if we want a default that represents the current value? Not sure we can even get to this point normally or what the consequences of 0 are, maybe better not to have it or is this proceeding assuming 0?

Reply:
I've moved to providing the config values directly in source for the time-being so we should be able to guarantee a price. I was thinking `0` was more for testing.